include/dali/core/tensor_layout.h

```diff
@@ -0,0 +1,341 @@
+DALI_HOST_DEV
+inline TensorLayout operator+(const TensorLayout &a, const TensorLayout &b) {
+  assert(a.size() + b.size() < TensorLayout::max_ndim);
+  TensorLayout result = a;
+  int i, j;
+  for (i = result.ndim(), j = 0; i < TensorLayout::max_ndim && j < b.ndim(); i++, j++)
+    result[i] = b[j];
+  result.size_ = i;
+  result[i] = 0;
+  return result;
+}
+
+static_assert(sizeof(TensorLayout) == 16, "Tensor layout size should be exactly 16B");
+
+#define DEFINE_TENSOR_LAYOUT_COMPARISON(op)                             \
+inline bool operator op(const TensorLayout &tl, const std::string &s) { \
+  return std::strcmp(tl.c_str(), s.c_str()) op 0;                       \
+}                                                                       \
+DALI_HOST_DEV                                                           \
+constexpr bool operator op(const TensorLayout &tl, const char *s) {     \
+  return tl.compare(s) op 0;                                            \
+}                                                                       \
+inline bool operator op(const std::string &s, const TensorLayout &tl) { \
+  return std::strcmp(s.c_str(), tl.c_str()) op 0;                       \
+}                                                                       \
+DALI_HOST_DEV                                                           \
+constexpr bool operator op(const char *s, const TensorLayout &tl) {     \
+  return TensorLayout(s).compare(tl) op 0;                              \
+}
+
+DEFINE_TENSOR_LAYOUT_COMPARISON(<)  // NOLINT
+DEFINE_TENSOR_LAYOUT_COMPARISON(>)  // NOLINT
+DEFINE_TENSOR_LAYOUT_COMPARISON(<=)
+DEFINE_TENSOR_LAYOUT_COMPARISON(>=)
+DEFINE_TENSOR_LAYOUT_COMPARISON(==)
+DEFINE_TENSOR_LAYOUT_COMPARISON(!=)
+
+/** Provides basic functions for querying TensorLayout properties */
+struct LayoutInfo {
+  DALI_HOST_DEV
+  static int HasSampleDim(const TensorLayout &tl) {
+    // sample dim may only appear at outermost level
+    return !tl.empty() && tl[0] == 'N';
+  }
+  DALI_HOST_DEV
+  static int DimIndex(const TensorLayout &tl, char dim_symbol) {
+    return tl.find(dim_symbol);
+  }
+};
+
+/** @brief Provides functions for querying TensorLayout properties,
+ *         assuming that the layout describes an image
+ */
+struct ImageLayoutInfo : LayoutInfo {
+  DALI_HOST_DEV
+  static int NumSpatialDims(const TensorLayout &tl) {
+    int s = 0;
+    for (int i = 0; i < tl.ndim(); i++) {
+      switch (tl[i]) {
+      case 'D':
+      case 'H':
+      case 'W':
+        s++;
+        break;
+      default:
+        break;
+      }
+    }
+    return s;
+  }
+  DALI_HOST_DEV
+  static int ChannelDimIndex(const TensorLayout &tl) {
+    return DimIndex(tl, 'C');
+  }
+  DALI_HOST_DEV
+  static int HasChannel(const TensorLayout &tl) {
+    return ChannelDimIndex(tl) >= 0;
+  }
+  DALI_HOST_DEV
+  static int Is2D(const TensorLayout &tl) {
+    return NumSpatialDims(tl) == 2;
+  }
+  DALI_HOST_DEV
+  static int Is3D(const TensorLayout &tl) {
```

Comment:
Why those return an int while the latter three return a bool?

Reply:
Nice catch.